pkg/frontend/querymiddleware/split_and_cache.go

```diff
@@ -365,23 +375,59 @@ func (s *splitAndCacheMiddleware) fetchCacheExtents(ctx context.Context, keys []
 			continue
 		}
 
-		extents[keyIdx] = resp.Extents
+		extents[keyIdx] = make([]Extent, 0, len(resp.Extents))
+
+		// Filter out extents that are outside TTL.
+		for ix := range resp.Extents {
+			if resp.Extents[ix].QueryTime == 0 {
+				// If we don't know the query time, it's too old.
+				extentsOutOfTTL++
+				continue
+			}
+
+			usedTTL := ttl
+			if oooWindow > 0 && extentWithinOOOWindow(&resp.Extents[ix], now, oooWindow) {
+				usedTTL = ttlForExtentsInOOOWindow
+			}
+			if resp.Extents[ix].QueryTime < now.UnixMilli()-usedTTL.Milliseconds() {
+				extentsOutOfTTL++
+				continue
+			}
+
+			extents[keyIdx] = append(extents[keyIdx], resp.Extents[ix])
+		}
+
+		if len(extents[keyIdx]) == 0 {
+			extents[keyIdx] = nil
+		}
+
 		returnedBytes += len(foundData)
 	}
 
 	spanLog.LogKV("requested keys", len(hashedKeys))
 	spanLog.LogKV("found keys", len(founds))
 	spanLog.LogKV("returned bytes", returnedBytes)
+	spanLog.LogKV("extents filtered out due to ttl", extentsOutOfTTL)
 
 	return extents
 }
 
+func (s *splitAndCacheMiddleware) getCacheOptions(tenantIDs []string) (ttl, ttlInOOO, oooWindow time.Duration) {
+	ttl = validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, s.limits.ResultsCacheTTL)
+	ttlInOOO = validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, s.limits.ResultsCacheTTLForOutOfOrderTimeWindow)
+	oooWindow = validation.MaxDurationPerTenant(tenantIDs, s.limits.OutOfOrderTimeWindow)
+	return
+}
+
 // storeCacheExtents stores the extents for given key in the cache.
 func (s *splitAndCacheMiddleware) storeCacheExtents(key string, tenantIDs []string, extents []Extent) {
```

Comment:
I just realised: we have some logic checking `len(extents) > 0` but what's the purpose of caching no extents? Should we just have a circuit breaker at the beginning of this function to skip the function execution if there are no extents?